bootstrap.rkt

```diff
@@ -0,0 +1,476 @@
+                                                                                     keyword
+                                                                                     '"did you mean (unquote symbol)?"
+                                                                                     '"did you mean to add the symbol to the keyword list?")
+                                                                                    stx)
+                                                                                  stx))))]
+                                                           [(fancy-unquote ,x)
+                                                            (if (identifier? x)
+                                                                (pure (generate-quasiquote
+                                                                        (let [,x ,scrutinee-name]
+                                                                          ,rhs)
+                                                                        stx))
+                                                                (syntax-error (list-syntax
+                                                                                ('"syntax-case: the syntax for binding values is (unquote x), found"
+                                                                                 pat
+                                                                                 '"instead")
+                                                                                stx)
+                                                                              stx))]
+                                                           [((fancy-unquote ,x) fancy-...)
+                                                            (if (identifier? x)
+                                                                (pure (generate-quasiquote
+                                                                        (let [,x ,scrutinee-name]
+                                                                          ,rhs)
+                                                                        stx))
+                                                                (syntax-error (list-syntax
+                                                                                ('"syntax-case: the syntax for binding lists is (,x ...), found"
+                                                                                 pat
+                                                                                 '"instead")
+                                                                                stx)
+                                                                              stx))]
+                                                           [(,pat-head ,pat-tail ...)
+                                                            (>>= (make-temporary 'tail)
```

Comment:
I think I might not have the right approach to writing macros. I don't like writing my recursive calls by generating code which calls my macro again, because that requires encoding all the information required by that recursive call in a syntax object. I find it much clearer to write recursive functions like `fancy-case` and `fancy-cases` which simply take ordinary values as input and return ordinary values as output, and then to have my macro call those functions in order to produce the final syntactic object. Unfortunately, I have found that this approach has a disadvantage: those helper functions all run in the same macro context, and so all the identifiers they manipulate have the same scope! Hygiene is thus not helping me here, and I have to call `make-temporary` (the Klister version of `gensym`). Is there a third alternative?